core/arch/arm/kernel/secure_partition.c

```diff
@@ -68,6 +85,22 @@ struct sp_session *sp_get_session(uint32_t session_id)
 	return NULL;
 }
 
+uint32_t sp_get_all_partitions(struct ffa_partition_info *fpi,
+			       size_t max_elements)
+{
+	struct sp_session *s = NULL;
+	uint32_t count = 0;
+
+	TAILQ_FOREACH(s, &open_sp_sessions, link) {
+		if (count < max_elements) {
```

Comment:
Generic comment: caller of FFA_PARTITION_INFO_GET may not see some partitions if it provides a too small `max_elements` value. This puts a constraint on the number of partitions and the rxtx buffer size. How to overcome that?

Reply:
@etienne-lms either use a similar interface to many GlobalPlatform APIs (`size_t *out_size` initialized to the client buffer size, then the function sets `*out_size` to the actual size needed). Or use a callback-based approach, thus providing one partition at a time.

Reply:
We discussed this internally and came up with the solution of returning a TEE_ERROR_OUT_OF_MEMORY. The caller needs to increase its buffer itself to make sure that there is enough space

Reply:
How about this?

```C
TEE_Result sp_get_all_partitions(struct ffa_partition_info *fpi,                
                                 size_t *elements)                              
{                                                                               
        size_t in_elements = *elements;                                         
        struct sp_session *s = NULL;                                            
        uint32_t count = 0;                                                     
                                                                                
        TAILQ_FOREACH(s, &open_sp_sessions, link) {                             
                if (count < in_elements) {                                      
                        spmc_fill_partition_entry(fpi, s->endpoint_id);         
                        fpi++;                                                  
                }                                                               
                count++;                                                        
        }                                                                       
                                                                                
        *elements = count;                                                      
        if (count > in_elements)                                               
                return TEE_ERROR_SHORT_BUFFER;                                  
                                                                                
        return TEE_SUCCESS;                                                     
}                                                                               
```
This is a common pattern, see for instance `get_devices()` in `core/pta/device.c`.

Reply:
Done

Reply:
note regarding jerome's code: when returning TEE_ERROR_SHORT_BUFFER as GP, the output buffer is not expected to be filled. So maybe return only the count when returning with TEE_ERROR_SHORT_BUFFER, and init/fill `fpi` only if returning TEE_SUCCESS.